src/lib.rs

```diff
@@ -1481,3 +1482,30 @@ macro_rules! impl_tuple_zeroable {
 }
 
 impl_tuple_zeroable!(A, B, C, D, E, F, G, H, I, J);
+
+/// Trait for creating a [PinInitialized](PinInit) wrapper containing `T`.
+pub trait TryPinInitWrapper<T: ?Sized> {
+    /// Create an pin-initializer for a [`Self`] containing `T` form the `value` initializer.
+    fn try_pin_init<E>(value: impl PinInit<T, E>) -> impl PinInit<Self, E>;
+}
+
+impl<T: ?Sized> TryPinInitWrapper<T> for UnsafeCell<T> {
+    fn try_pin_init<E>(value: impl PinInit<T, E>) -> impl PinInit<Self, E> {
+        // SAFETY:
+        //   - In case of an error in `value` the error is returned, otherwise `slot` is fully
+        //     initialized, since `self.value` is initialized and `_pin` is a zero sized type.
+        //   - The `Pin` invariants of `self.value` are upheld, since no moving occurs.
+        unsafe { pin_init_from_closure(move |slot| value.__pinned_init(Self::raw_get(slot))) }
+    }
+}
+
+#[cfg(feature = "unsafe-pinned")]
+impl<T: ?Sized> TryPinInitWrapper<T> for core::pin::UnsafePinned<T> {
```

Comment:
can you separate the `unsafe-pinned` feature & adding of this impl from adding the `TryPinInitWrapper`/`PinInitable`?